Index: scite/gtk/SciTEGTK.cxx
===================================================================
--- scite/gtk/SciTEGTK.cxx	(版本 49)
+++ scite/gtk/SciTEGTK.cxx	(工作副本)
@@ -1697,7 +1697,7 @@
 		OutputAppendString(buf);
 	} else if (count == 0) {
 		SString sExitMessage(WEXITSTATUS(exitStatus));
-		sExitMessage.insert(0, ">Exit code: ");
+		sExitMessage.insert(0, "\n>Exit code: ");//!-[change exit]
 		if (WIFSIGNALED(exitStatus)) {
 			SString sSignal(WTERMSIG(exitStatus));
 			sSignal.insert(0, " Signal: ");
Index: scite/src/Extender.h
===================================================================
--- scite/src/Extender.h	(版本 49)
+++ scite/src/Extender.h	(工作副本)
@@ -65,7 +65,7 @@
 	virtual bool OnMarginClick() { return false; }
 	virtual bool OnMacro(const char *, const char *) { return false; }
 	virtual bool OnUserListSelection(int, const char *) { return false; }
-
+	virtual bool OnMenuCommand(int, int) { return false; } //!-add-[OnMenuCommand]
 	virtual bool SendProperty(const char *) { return false; }
 
 	virtual bool OnKey(int, int) { return false; }
Index: scite/src/IFaceTable.cxx
===================================================================
--- scite/src/IFaceTable.cxx	(版本 49)
+++ scite/src/IFaceTable.cxx	(工作副本)
@@ -166,6 +166,10 @@
 	{"IDM_COMPLETEWORD",234},
 	{"IDM_COPY",204},
 	{"IDM_COPYASRTF",245},
+	{"IDM_COPYFILEDIR",370},
+	{"IDM_COPYFILEEXT",373},
+	{"IDM_COPYFILENAMEEXT",371},
+	{"IDM_COPYFILENAME",372},
 	{"IDM_COPYPATH",118},
 	{"IDM_CUT",203},
 	{"IDM_DUPLICATE",250},
@@ -265,6 +269,12 @@
 	{"IDM_STATUSWIN",353},
 	{"IDM_STOPEXECUTE",304},
 	{"IDM_STREAM_COMMENT",244},
+	{"IDM_SVN_ADD",362},
+	{"IDM_SVN_COMMIT",361},
+	{"IDM_SVN_DIFF",364},
+	{"IDM_SVN_REVERT",363},
+	{"IDM_SVN_SHOWLOG",365},
+	{"IDM_SVN_UPDATE",360},
 	{"IDM_SWITCHPANE",421},
 	{"IDM_TABSIZE",440},
 	{"IDM_TABWIN",354},
Index: scite/src/LuaExtension.cxx
===================================================================
--- scite/src/LuaExtension.cxx	(版本 49)
+++ scite/src/LuaExtension.cxx	(工作副本)
@@ -762,6 +762,23 @@
 	return handled;
 }
 
+//!-start-[OnMenuCommand]
+static bool CallNamedFunction(const char *name, int numberArg, int numberArg2) {
+	bool handled = false;
+	if (luaState) {
+		lua_getglobal(luaState, name);
+		if (lua_isfunction(luaState, -1)) {
+			lua_pushnumber(luaState, numberArg);
+			lua_pushnumber(luaState, numberArg2);
+			handled = call_function(luaState, 2);
+		} else {
+			lua_pop(luaState, 1);
+		}
+	}
+	return handled;
+}
+//!-end-[OnMenuCommand]
+
 static bool CallNamedFunction(const char *name) {
 	bool handled = false;
 	if (luaState) {
@@ -1620,6 +1637,12 @@
 	return CallNamedFunction("OnSavePointLeft");
 }
 
+//!-start-[OnMenuCommand]
+bool LuaExtension::OnMenuCommand(int cmd, int source) {
+	return CallNamedFunction("OnMenuCommand", cmd, source);
+}
+//!-end-[OnMenuCommand]
+
 // Similar to StyleContext class in Scintilla
 struct StylingContext {
 	unsigned int startPos;
Index: scite/src/LuaExtension.h
===================================================================
--- scite/src/LuaExtension.h	(版本 49)
+++ scite/src/LuaExtension.h	(工作副本)
@@ -36,6 +36,7 @@
 	virtual bool OnUpdateUI();
 	virtual bool OnMarginClick();
 	virtual bool OnUserListSelection(int listType, const char *selection);
+	virtual bool OnMenuCommand(int, int); //!-add-[OnMenuCommand]
 	virtual bool OnKey(int keyval, int modifiers);
 	virtual bool OnDwellStart(int pos, const char *word);
 	virtual bool OnClose(const char *filename);
Index: scite/src/MultiplexExtension.cxx
===================================================================
--- scite/src/MultiplexExtension.cxx	(版本 49)
+++ scite/src/MultiplexExtension.cxx	(工作副本)
@@ -247,3 +247,12 @@
 	return false;
 }
 
+//!-start-[OnMenuCommand]
+bool MultiplexExtension::OnMenuCommand(int cmd, int source) {
+	bool handled = false;
+	for (int i=0; i<extensionCount && !handled; ++i)
+                if (extensions[i]->OnMenuCommand(cmd,source))
+                        handled = true;
+	return handled;
+}
+//!-end-[OnMenuCommand]
Index: scite/src/MultiplexExtension.h
===================================================================
--- scite/src/MultiplexExtension.h	(版本 49)
+++ scite/src/MultiplexExtension.h	(工作副本)
@@ -67,7 +67,7 @@
 	virtual bool OnMarginClick();
 	virtual bool OnMacro(const char *, const char *);
 	virtual bool OnUserListSelection(int, const char *);
-
+	virtual bool OnMenuCommand(int, int); //!-add-[OnMenuCommand]
 	virtual bool SendProperty(const char *);
 
 	virtual bool OnKey(int, int);
Index: scite/src/SciTE.h
===================================================================
--- scite/src/SciTE.h	(版本 49)
+++ scite/src/SciTE.h	(工作副本)
@@ -14,12 +14,12 @@
 #define IDM_TOOLS			1100
 #define IDM_BUFFER			1200
 #define IDM_IMPORT			1300
-#define IDM_LANGUAGE			1400
+#define IDM_LANGUAGE		1400
 
 // File
-#define IDM_NEW			101
+#define IDM_NEW				101
 #define IDM_OPEN			102
-#define IDM_OPENSELECTED		103
+#define IDM_OPENSELECTED	103
 #define IDM_REVERT			104
 #define IDM_CLOSE			105
 #define IDM_SAVE			106
@@ -31,7 +31,7 @@
 #define IDM_SAVEASTEX		115
 #define IDM_SAVEACOPY		116
 #define IDM_SAVEASXML		117
-#define IDM_COPYPATH			118
+#define IDM_COPYPATH		118
 #define IDM_MRU_SEP			120
 #define IDM_PRINTSETUP		130
 #define IDM_PRINT			131
@@ -121,7 +121,7 @@
 #define IDM_MACROPLAY		313
 #define IDM_MACROLIST		314
 
-#define IDM_ACTIVATE			320
+#define IDM_ACTIVATE		320
 
 #define IDM_SRCWIN			350
 #define IDM_RUNWIN			351
@@ -129,6 +129,22 @@
 #define IDM_STATUSWIN		353
 #define IDM_TABWIN			354
 
+//!-[add SVN menu IDs]
+#define IDM_SVN_UPDATE      360
+#define IDM_SVN_COMMIT		361
+#define IDM_SVN_ADD			362
+#define IDM_SVN_REVERT		363
+#define IDM_SVN_DIFF		364
+#define IDM_SVN_SHOWLOG		365
+//!-[add SVN menu IDs]
+
+//!-[add copy filepath]
+#define IDM_COPYFILEDIR     370
+#define IDM_COPYFILENAMEEXT 371
+#define IDM_COPYFILENAME    372
+#define IDM_COPYFILEEXT     373
+//!-[add copy filepath]
+
 // Options
 #define IDM_SPLITVERTICAL		401
 #define IDM_VIEWSPACE		402
Index: scite/src/SciTEBase.cxx
===================================================================
--- scite/src/SciTEBase.cxx	(版本 49)
+++ scite/src/SciTEBase.cxx	(工作副本)
@@ -76,9 +76,11 @@
 #include "JobQueue.h"
 #include "SciTEBase.h"
 
+#define _MAX_EXTENSION_RECURSIVE_CALL 100 //!-add-[OnSendEditor][OnMenuCommand]
+
 // Contributor names are in UTF-8
 const char *contributors[] = {
-			"Kim Long",
+			"Kimha Long",
             "Atsuo Ishimoto",
             "Mark Hammond",
             "Francois Le Coguiec",
@@ -457,6 +459,8 @@
 	propsStatus.superPS = &props;
 
 	needReadProperties = false;
+	
+	OnMenuCommandCallsCount = 0;	//!-add-[OnMenuCommand]
 }
 
 SciTEBase::~SciTEBase() {
@@ -3318,6 +3322,14 @@
 }
 
 void SciTEBase::MenuCommand(int cmdID, int source) {
+	//!-start-[OnMenuCommand]
+	if (extender && OnMenuCommandCallsCount < _MAX_EXTENSION_RECURSIVE_CALL) {
+		OnMenuCommandCallsCount++;
+		bool result = extender->OnMenuCommand(cmdID,source);
+		OnMenuCommandCallsCount--;
+		if (result) return;
+	}
+	//!-end-[OnMenuCommand]
 	switch (cmdID) {
 	case IDM_NEW:
 		// For the New command, the "are you sure" question is always asked as this gives
Index: scite/src/SciTEBase.h
===================================================================
--- scite/src/SciTEBase.h	(版本 49)
+++ scite/src/SciTEBase.h	(工作副本)
@@ -889,6 +889,7 @@
 	bool iswordcharforsel(char ch);
 	bool isfilenamecharforsel(char ch);
 	bool islexerwordcharforsel(char ch);
+	int OnMenuCommandCallsCount; //!-add-[OnMenuCommand]
 public:
 
 	SciTEBase(Extension *ext = 0);
Index: scite/win32/SciTERes.rc
===================================================================
--- scite/win32/SciTERes.rc	(版本 49)
+++ scite/win32/SciTERes.rc	(工作副本)
@@ -155,6 +155,24 @@
         MENUITEM "&Build\tF7",                  IDM_BUILD
         MENUITEM "&Go\tF5",                     IDM_GO
         MENUITEM "&Stop Executing\tCtrl+Break", IDM_STOPEXECUTE
+		MENUITEM SEPARATOR
+        POPUP "&Subversion"
+        BEGIN
+            MENUITEM "&Update",                 IDM_SVN_UPDATE
+            MENUITEM "&Commit",                 IDM_SVN_COMMIT
+            MENUITEM "&Add...",                 IDM_SVN_ADD
+            MENUITEM "&Revert To...",           IDM_SVN_REVERT
+            MENUITEM "&Diff...",                IDM_SVN_DIFF
+            MENUITEM "&Show Log",               IDM_SVN_SHOWLOG
+        END
+		POPUP "&Copy To Clipboard"
+        BEGIN
+            MENUITEM "File Full Path",          IDM_COPYPATH
+            MENUITEM "File Directory",          IDM_COPYFILEDIR
+            MENUITEM "File Name With Ext",      IDM_COPYFILENAMEEXT
+            MENUITEM "File Name",               IDM_COPYFILENAME
+            MENUITEM "File Ext",            	IDM_COPYFILEEXT
+        END
         MENUITEM SEPARATOR
         MENUITEM "&Next Message\tF4",           IDM_NEXTMSG
         MENUITEM "&Previous Message\tShift+F4", IDM_PREVMSG
Index: scite/win32/SciTEWin.cxx
===================================================================
--- scite/win32/SciTEWin.cxx	(版本 49)
+++ scite/win32/SciTEWin.cxx	(工作副本)
@@ -888,7 +888,7 @@
 		}
 		::GetExitCodeProcess(pi.hProcess, &exitcode);
 		SString sExitMessage(exitcode);
-		sExitMessage.insert(0, ">Exit code: ");
+		sExitMessage.insert(0, "\n>Exit code: ");//!-[change exit]
 		if (jobQueue.TimeCommands()) {
 			sExitMessage += "    Time: ";
 			sExitMessage += SString(commandTime.Duration(), 3);
Index: scite/win32/SciTEWinBar.cxx
===================================================================
--- scite/win32/SciTEWinBar.cxx	(版本 49)
+++ scite/win32/SciTEWinBar.cxx	(工作副本)
@@ -132,9 +132,16 @@
 			AddToPopUp("Save", IDM_SAVE, true);
 			AddToPopUp("Save As", IDM_SAVEAS, true);
 			AddToPopUp("");
-
-			bool bAddSeparator = false;
-			for (int item = 0; item < toolMax; item++) {
+			AddToPopUp("Copy File Full Path", IDM_COPYPATH, true);
+			AddToPopUp("Copy File Directory", IDM_COPYFILEDIR, true);
+			AddToPopUp("Copy File Name With Ext", IDM_COPYFILENAMEEXT, true);
+			AddToPopUp("Copy File Name", IDM_COPYFILENAME, true);
+			AddToPopUp("Copy File Ext", IDM_COPYFILEEXT, true);
+			
+			
+			//bool bAddSeparator = false;
+			//!-[change NM_RCLICK]
+			/*for (int item = 0; item < toolMax; item++) {
 				int itemID = IDM_TOOLS + item;
 				SString prefix = "command.name.";
 				prefix += SString(item);
@@ -147,10 +154,10 @@
 					AddToPopUp(sMenuItem.c_str(), itemID, true);
 					bAddSeparator = true;
 				}
-			}
+			}*/
 
-			if (bAddSeparator)
-				AddToPopUp("");
+			//if (bAddSeparator)
+			AddToPopUp("");
 
 			AddToPopUp("Print", IDM_PRINT, true);
 			popup.Show(ptCursor, wSciTE);
Index: scite/win32/SingleThreadExtension.cxx
===================================================================
--- scite/win32/SingleThreadExtension.cxx	(版本 49)
+++ scite/win32/SingleThreadExtension.cxx	(工作副本)
@@ -159,3 +159,9 @@
 bool SingleThreadExtension::OnClose(const char *filename) {
 	return ext->OnClose(filename);
 }
+
+//!-start-[OnMenuCommand]
+bool SingleThreadExtension::OnMenuCommand(int cmd, int source) {
+	return ext->OnMenuCommand(cmd, source);
+}
+//!-end-[OnMenuCommand]
\ No newline at end of file
Index: scite/win32/SingleThreadExtension.h
===================================================================
--- scite/win32/SingleThreadExtension.h	(版本 49)
+++ scite/win32/SingleThreadExtension.h	(工作副本)
@@ -45,6 +45,7 @@
 	virtual bool OnKey(int, int);
 	virtual bool OnDwellStart(int, const char *);
 	virtual bool OnClose(const char *);
+	virtual bool OnMenuCommand(int, int); //!-add-[OnMenuCommand]
 
 protected:
 	Extension *ext;
